rfcs/0000-light-rail-trip-model.md

```diff
@@ -0,0 +1,211 @@
+
+The `com.mbta.ctd.glides.trips_updated` event schema will receive a version bump, modifying the `TripKey` datatype. `TripKey` will be the same for both added and scheduled trips, in either case being represented as an object with `"serviceDate"` and `"tripId"` keys, where the `tripId` is sourced from the `trip_id` column of `scheduled_trips` for a scheduled trip, or the auto-generated Glides ID for an added trip.
+
+## Glides train - trip assignment events
+
+A new `com.mbta.ctd.glides.vehicle_trip_assignment` event will be established. The exact schema will be defined via the pull request review procedure on the [`mbta/schemas`](https://github.com/mbta/schemas/) repository, but there are some basic criteria that can be established. Events will at least have a `tripId` and `vehicleId` key. The `tripId` will be the scheduled or added trip ID from Glides, and the `vehicleId` will correspond to the ID of the vehicle in the GTFS-rt feed. Each event will represent the assignment of the given `vehicleId` to the given `tripId`. Either a variation of this event or a separate event will exist for unassigning a vehicle from a trip.
+
+The Glides application will be primarily responsible for establishing the train - trip matching. While exact details of the logic are beyond the scope of this RFC, factors that are likely to be weighed include:
+
+1. Cars entered for trips on trainsheets
+1. Additional mid-line edits to trips by Glides users, however those are represented
+1. Train movement (for instance, marking a train as no longer assigned to a trip when it reaches that trip's endpoint, or assigning it to a different trip if appropriate)
+1. Train location (for instance, marking a train as no longer assigned to a trip if its current location is no longer on the path from the start point to the end point, including being headed in the correct direction)
+1. AVI code (indicating train destination)
+
+While there are not hard requirements around this, Glides SHOULD generally try to produce trip assignments for any vehicle tracking in the realtime data that is operating on the mainline (as opposed to in a yard). This is important for RTR to be able to make predictions without doing additional trip assignment logic (any train operating in service should be getting predictions, whether revenue train predictions in the public GTFS-rt feed or non-revenue predictions for in-station announcements about trains that won't stop). Additionally, it has benefits for operations (having a trip in the internal Glides data model provides a location to store any additional metadata, such as notes). For the time being, trip matches performed to provide a placeholder trip for a train not associated with a trip explicitly via the cars entered on a trainsheet will generally be ADDED trips that do not appear on the trainsheet interface. They will, however, generate corresponding `com.mbta.ctd.glides.trips_updated` events creating those added trips.
```

Comment:
I think this partially answers the above question: If an operator doesn't assign a train, the trip will be ADDED, but this still requires detecting when a train begins a new trip and assigning the correct information to it when an operator fails to input the information, which I think requires the process described above.

Reply:
I think my above comments reply to this somewhat, and I'll also add that when a trip in Glides is scheduled, the act of the inspector at the departing terminal entering the car numbers also generally enters the car numbers for the corresponding return trip, so in that case Glides will easily be able to detect that the previous trip has completed and the new trip with the same car numbers is now in progress.

Reply:
> the act of the inspector at the departing terminal entering the car numbers also generally enters the car numbers for the corresponding return trip, so in that case Glides will easily be able to detect that the previous trip has completed and the new trip with the same car numbers is now in progress.

I think this may at least partially answer my question above, although are we assuming we can reliably get inspectors to start entering car numbers? If not we still have the sad path to consider.

Reply:
Yes, this still relies on inspector input, although in terms of the quality of the data being entered, the problems are more around making sure that the data is entered in advance of the initial departure. In the BC - Government Center example, if a train departs BC and the inspector doesn't enter the car numbers and so on until the train is actually leaving, that's bad for predictions from BC but in terms of figuring out that that train is then going to turn around and make the following trip from Government Center back, that still gives us plenty of lead time.